include/cifer/quadratic/sgp.h

```diff
@@ -0,0 +1,180 @@
+/* *
+ * Copyright (C) 2018 XLAB d.o.o.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of either:
+ *
+ *     * the GNU Lesser General Public License as published by the Free
+ *     Software Foundation; either version 3 of the License, or (at your
+ *     option) any later version.
+ *
+ * or
+ *
+ *     * the GNU General Public License as published by the Free Software
+ *     Foundation; either version 2 of the License, or (at your option) any
+ *     later version.
+ *
+ * or both in parallel, as here.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+#ifndef CIFER_SGP_H
+#define CIFER_SGP_H
+
+#include "cifer/data/vec.h"
+#include "cifer/internal/errors.h"
+#include "cifer/data/vec_curve.h"
+
+/**
+ * \file
+ * \ingroup quadratic
+ * \brief SGP scheme.
+ */
+
+/**
+ * cfe_sgp represents a scheme for quadratic multi-variate polynomials
+// based on  Sans, Gay and Pointcheval:
+// "Reading in the Dark: Classifying Encrypted Digits with
+// Functional Encryption", see https://eprint.iacr.org/2018/206.pdf.
+ */
+typedef struct cfe_sgp {
+    size_t n;
+    mpz_t bound;
+    mpz_t mod;
+    BIG_256_56 modBig;
+} cfe_sgp;
+
+/**
+ * cfe_sgp_sec_key represents a secret key for SGP scheme.
+ */
+typedef struct cfe_sgp_sec_key {
+    cfe_vec s;
+    cfe_vec t;
+} cfe_sgp_sec_key;
+
+/**
+ * cfe_sgp_sec_key represents a ciphertext in SGP scheme.
+ */
+typedef struct cfe_sgp_cipher {
+    ECP_BN254 g1MulGamma;
+    cfe_vec_G1 *a;
+    cfe_vec_G2 *b;
```

Comment:
I think you can change the pointers to vectors to just vectors in this struct.

Reply:
For a moment I thought so also, but it is really an array of vectors.